File: framework/headless/combobox.js
// headless-combobox: the behaviour module for this package's
// comboboxes — a [role=combobox] input paired with its
// [role=listbox] through aria-controls. The input owns focus through
// every interaction; the module owns the keyboard contract
// (ArrowDown/ArrowUp open and move aria-activedescendant with wrap,
// Home/End jump, Enter commits the active option, Escape closes and a
// second Escape clears the query, Tab closes and lets focus move),
// pointer picks, focus auto-open, outside-click close, and the static
// list's client-side filtering. The RPC debounce and the signal swap
// are the kernel's data-fui-rpc contract, not reimplemented here.
(function () {
  'use strict';
  const NAME = 'headless-combobox';
  const NS = window.__gofastr = window.__gofastr || {};
  if (NS.loadedModules && Object.prototype.hasOwnProperty.call(NS.loadedModules, NAME)) return;
  NS.loadedModules = NS.loadedModules || {};
  NS.loadedModules[NAME] = true;

  function within(root, sel) {
    const out = [];
    if (root.matches && root.matches(sel)) out.push(root);
    if (root.querySelectorAll) out.push.apply(out, root.querySelectorAll(sel));
    return out;
  }

  function inputOf(lb) {
    if (!lb.id) return null;
    return document.querySelector('[data-hui-combobox-input][aria-controls="' + cssEscape(lb.id) + '"]');
  }

  function cssEscape(s) {
    if (window.CSS && CSS.escape) return CSS.escape(s);
    let out = '';
    if (s.length > 0 && s.charCodeAt(0) >= 48 && s.charCodeAt(0) <= 57) {
      out = '\\3' + s.charAt(0) + ' ';
      s = s.slice(1);
    }
    return out + s.replace(/([!"#$%&'()*+,./:;<=>?@[\]^`{|}~])/g, '\\$1');
  }

  function highlight(lb, opt) {
    for (const o of lb.querySelectorAll('[role="option"].is-active')) o.classList.remove('is-active');
    if (opt) {
      opt.classList.add('is-active');
      const input = inputOf(lb);
      if (input) input.setAttribute('aria-activedescendant', opt.id || '');
    }
  }

  function say(lb, what) {
    const input = inputOf(lb);
    const root = input && input.closest('[data-hui-combobox]');
    const status = root && root.querySelector('[data-hui-combobox-status]');
    if (status) status.textContent = what || '';
  }

  function announce(lb) {
    // The count sentence travels on the listbox; {n} is written in
    // when the count is known. No sentence, no announcement.
    const fmt = lb.getAttribute('data-hui-combobox-count');
    if (!fmt) return;
    const n = lb.querySelectorAll('[role="option"]:not([hidden])').length;
    if (n === 0) {
      say(lb, lb.getAttribute('data-hui-combobox-no-results') || '');
      return;
    }
    say(lb, fmt.replace('{n}', String(n)));
  }

  function closeListbox(input, lb) {
    input.setAttribute('aria-expanded', 'false');
    input.setAttribute('aria-activedescendant', '');
    if (lb) {
      lb.setAttribute('hidden', '');
      for (const o of lb.querySelectorAll('[role="option"].is-active')) o.classList.remove('is-active');
    }
  }

  function openListbox(input, lb) {
    if (!lb) return;
    input.setAttribute('aria-expanded', 'true');
    lb.removeAttribute('hidden');
  }

  // sameOriginDest for the pre-boot pick path: a javascript: or data:
  // URL resolves to the opaque "null" origin and is refused.
  function sameOriginDest(u) {
    try { return new URL(String(u ?? ''), location.href).origin === location.origin; }
    catch (_) { return false; }
  }

  function pickOption(input, lb, opt) {
    if (!opt) return;
    const val = opt.getAttribute('data-value') || (opt.textContent || '').trim();
    input.value = val;
    input.dispatchEvent(new Event('change', { bubbles: true }));
    closeListbox(input, lb);
    // A picked option's destination rides data-fui-push-state (or the
    // anchor's own href): a same-origin pick navigates through the SPA
    // navigator, closing any enclosing widget first so the navigation
    // is not behind a backdrop.
    const dest = opt.getAttribute('data-fui-push-state') ||
      (opt.tagName === 'A' ? opt.getAttribute('href') : null);
    if (dest && (NS._originOK ? NS._originOK(dest) : sameOriginDest(dest))) {
      const widget = opt.closest('[data-fui-widget]');
      if (widget && NS.closeWidget) {
        try { NS.closeWidget(widget.getAttribute('data-fui-widget')); } catch (_) {}
      }
      if (NS.navigate) NS.navigate(dest);
      else location.href = dest;
    }
  }

  function options(lb) {
    return Array.from(lb.querySelectorAll(
      '[role="option"]:not([aria-disabled="true"]):not([hidden])'
    ));
  }

  document.addEventListener('keydown', function (e) {
    const input = e.target && e.target.closest && e.target.closest('[data-hui-combobox-input]');
    if (!input) return;
    const lbId = input.getAttribute('aria-controls');
    if (!lbId) return;
    const lb = document.getElementById(lbId);
    if (!lb) return;
    const opts = options(lb);
    const activeId = input.getAttribute('aria-activedescendant');
    const activeIdx = opts.findIndex(function (o) { return o.id === activeId; });
    const isOpen = input.getAttribute('aria-expanded') === 'true';

    switch (e.key) {
      case 'ArrowDown': {
        if (opts.length === 0) return;
        e.preventDefault();
        if (!isOpen) { openListbox(input, lb); highlight(lb, opts[0]); return; }
        highlight(lb, opts[(activeIdx + 1 + opts.length) % opts.length]);
        return;
      }
      case 'ArrowUp': {
        if (opts.length === 0) return;
        e.preventDefault();
        if (!isOpen) { openListbox(input, lb); highlight(lb, opts[opts.length - 1]); return; }
        highlight(lb, opts[(activeIdx - 1 + opts.length) % opts.length]);
        return;
      }
      case 'Home': {
        if (!isOpen || opts.length === 0) return;
        e.preventDefault();
        highlight(lb, opts[0]);
        return;
      }
      case 'End': {
        if (!isOpen || opts.length === 0) return;
        e.preventDefault();
        highlight(lb, opts[opts.length - 1]);
        return;
      }
      case 'Enter': {
        if (!isOpen || activeIdx < 0) return;
        e.preventDefault();
        pickOption(input, lb, opts[activeIdx]);
        return;
      }
      case 'Escape': {
        if (isOpen) { e.preventDefault(); closeListbox(input, lb); return; }
        if (input.value) {
          e.preventDefault();
          input.value = '';
          input.dispatchEvent(new Event('input', { bubbles: true }));
        }
        return;
      }
      case 'Tab': {
        if (isOpen) closeListbox(input, lb);
        return;
      }
    }
  });

  // Click-to-pick, delegated so island-swapped options are covered.
  document.addEventListener('click', function (e) {
    const opt = e.target && e.target.closest && e.target.closest('[role="option"]');
    if (!opt || opt.getAttribute('aria-disabled') === 'true') return;
    const lb = opt.closest('[data-hui-combobox-listbox]');
    if (!lb || !lb.id) return;
    const input = inputOf(lb);
    if (!input) return;
    e.preventDefault();
    pickOption(input, lb, opt);
  });

  // Focus auto-open: a refocused input whose listbox already holds
  // options reopens so the reader can continue.
  document.addEventListener('focusin', function (e) {
    const input = e.target && e.target.closest && e.target.closest('[data-hui-combobox-input]');
    if (!input) return;
    const lbId = input.getAttribute('aria-controls');
    const lb = lbId ? document.getElementById(lbId) : null;
    if (lb && lb.querySelector('[role="option"]')) openListbox(input, lb);
  });

  // Outside-click closes any open combobox.
  document.addEventListener('click', function (e) {
    for (const input of document.querySelectorAll('[data-hui-combobox-input][aria-expanded="true"]')) {
      const lbId = input.getAttribute('aria-controls');
      const lb = lbId ? document.getElementById(lbId) : null;
      if (input.contains(e.target) || (lb && lb.contains(e.target))) continue;
      closeListbox(input, lb);
    }
  });

  // The island dialect's loading word: the input event fired the RPC
  // (the kernel owns the debounce and the swap); the status region
  // says the word until the swap lands and the arrival pass announces
  // the count. The word rides the loader carrier's hook.
  document.addEventListener('input', function (e) {
    const loader = e.target && e.target.closest && e.target.closest('[data-hui-combobox-loader]');
    if (loader) {
      const root = loader.closest('[data-hui-combobox]');
      const status = root && root.querySelector('[data-hui-combobox-status]');
      const word = loader.getAttribute('data-hui-combobox-loading') || '';
      if (status && word) status.textContent = word;
    }
  }, true);

  // Static-option filtering: hide non-matches, show all when the query
  // clears, highlight the first match, and announce through the count
  // sentence.
  document.addEventListener('input', function (e) {
    const input = e.target && e.target.closest && e.target.closest('[data-hui-combobox-input]');
    if (!input) return;
    const lbId = input.getAttribute('aria-controls');
    const lb = lbId ? document.getElementById(lbId) : null;
    if (!lb || !lb.hasAttribute('data-hui-combobox-static')) return;
    const q = (input.value || '').toLowerCase().trim();
    let firstVisible = null;
    let anyVisible = false;
    lb.querySelectorAll('[role="option"]').forEach(function (opt) {
      const match = !q || (opt.textContent || '').toLowerCase().indexOf(q) !== -1;
      opt.hidden = !match;
      if (match) { anyVisible = true; if (!firstVisible) firstVisible = opt; }
    });
    if (anyVisible) { openListbox(input, lb); highlight(lb, firstVisible); }
    else input.setAttribute('aria-activedescendant', '');
    announce(lb);
  });

  // The arrival pass: announce the static list's count, and close any
  // listbox whose island region was swapped empty.
  function scan(root) {
    const scope = root && root.querySelectorAll ? root : document;
    for (const lb of within(scope, '[data-hui-combobox-listbox]')) {
      if (lb.hasAttribute('data-hui-combobox-static')) announce(lb);
    }
  }

  scan(document);
  NS._moduleScanners = NS._moduleScanners || {};
  NS._moduleScanners[NAME] = scan;
})();

File: framework/headless/combobox.go
package headless

import (
	"strconv"

	"github.com/DonaldMurillo/gofastr/core-ui/html"
	"github.com/DonaldMurillo/gofastr/core-ui/urlsafe"
	"github.com/DonaldMurillo/gofastr/core/render"
)

// The combobox: an input that owns a listbox of options. The input
// keeps focus through every interaction (the ARIA pattern's rule); the
// listbox is linked by aria-controls, the active option by
// aria-activedescendant, and the open state by aria-expanded — all
// rendered server-side so a reader with no script gets a labelled
// search form whose results are real content. The module owns the
// keyboard contract and the pick; the RPC debouncing and the signal
// swap are the kernel's data-fui-rpc contract, rendered here beside
// the form's no-script destination.

// Combobox parts.
const (
	PartComboboxForm    Part = "combobox-form"
	PartComboboxInput   Part = "combobox-input"
	PartComboboxListbox Part = "combobox-listbox"
	PartComboboxOption  Part = "combobox-option"
	PartComboboxStatus  Part = "combobox-status"
)

// ComboboxOption is one row in the listbox.
type ComboboxOption struct {
	// ID is the option's stable id. Empty takes "<listboxID>-opt-<n>".
	ID string
	// Value is what a pick writes into the input. Empty takes Label.
	Value string
	// Label is the option's visible text. Required.
	Label string
	// Meta is secondary text beside the label (a path, a kind).
	Meta string
	// Href turns the option into an anchor: picking it navigates.
	// Unsafe schemes drop the navigation affordance entirely (the
	// pick still fills the input).
	Href string
	// Disabled removes the option from keyboard navigation.
	Disabled bool
}

// ComboboxProps configures the combobox.
type ComboboxProps struct {
	// ID is the input's element id. Required; the listbox takes
	// "<ID>-listbox".
	ID string
	// Name is the form-submit name on the input. Required.
	Name string
	// Label names the input. Required.
	Label string

	// Placeholder for the input.
	Placeholder string

	// Island is the typed in-page results contract: the endpoint that
	// re-renders the listbox and the signal the region is bound to.
	// Optional — a static Options list needs no island; the no-script
	// path is the same-origin GET form below either way.
	Island *Island

	// NoScriptAction is the form's action URL, the no-script
	// destination: same-origin, a GET that submits the query. Required
	// when Island is set (a reader without script must still reach the
	// results); refused when it is "#".
	NoScriptAction string

	// Options is a static list the module filters client-side. Takes
	// precedence over the Island (no round-trip fires).
	Options []ComboboxOption

	// DebounceMS bounds the input debounce. Zero takes 250; negative
	// is refused.
	DebounceMS int

	ExtraAttrs html.Attrs

	// Parts: attrs on the root, the input and the listbox. The
	// listbox's content is the Options' — not fillable.
	Parts   Parts
	Strings *Strings
}

// Combobox renders the search input with its listbox.
func Combobox(p ComboboxProps, s Classes) render.HTML {
	if p.ID == "" {
		panic("headless: Combobox requires ID — the listbox is paired to the input by it")
	}
	if p.Name == "" {
		panic("headless: Combobox requires Name — the no-script form submits the query under it")
	}
	checkLabel("Combobox", "Label", p.Label)
	hasStatic := len(p.Options) > 0
	if p.Island == nil && !hasStatic {
		panic("headless: Combobox requires Island or Options — a combobox whose results come from nowhere is a text field")
	}
	if p.Island != nil {
		p.Island.check()
		if p.NoScriptAction == "" {
			panic("headless: Combobox requires NoScriptAction when Island is set — a reader without script must still reach the results")
		}
		if p.NoScriptAction == "#" {
			panic("headless: Combobox NoScriptAction must be a real same-origin destination, not #")
		}
		checkSameOrigin("Combobox", "NoScriptAction", p.NoScriptAction)
	}
	if p.DebounceMS < 0 {
		panic("headless: Combobox DebounceMS " + strconv.Itoa(p.DebounceMS) + " is negative — a negative debounce fires before the keystroke")
	}
	w := p.Strings.Resolve()
	debounce := p.DebounceMS
	if debounce == 0 {
		debounce = 250
	}
	listboxID := p.ID + "-listbox"

	// Option ids: caller's, checked for duplicates, or derived.
	ids := make([]string, 0, len(p.Options))
	for i := range p.Options {
		if p.Options[i].ID == "" {
			p.Options[i].ID = listboxID + "-opt-" + strconv.Itoa(i)
		}
		checkFragmentID("Combobox option", "ID", p.Options[i].ID)
		ids = append(ids, p.Options[i].ID)
	}
	checkNoDuplicateIDs("Combobox", ids)

	b := p.Parts.Box(s)
	inputAttrs := Attrs(map[string]string{
		"type":              "text",
		"id":                p.ID,
		"name":              p.Name,
		"role":              "combobox",
		"aria-autocomplete": "list",
		"aria-controls":     listboxID,
		"aria-expanded":     "false",
		"autocomplete":      "off",
		"spellcheck":        "false",
	})
	Mark(inputAttrs, "data-hui-combobox-input")
	if p.Placeholder != "" {
		inputAttrs["placeholder"] = scrubControlBytes(p.Placeholder)
	}

	// The carrier is a DIV, not a FORM: the HTML parser drops a nested
	// <form> open tag but honors its </form>, closing any host form the
	// combobox is embedded in. The RPC trigger attributes work on any
	// carrier; the real form (the no-script GET) wraps the whole
	// combobox instead.
	carrierAttrs := Attrs(map[string]string{})
	if p.Island != nil && !hasStatic {
		carrierAttrs["data-fui-rpc"] = p.Island.Endpoint
		carrierAttrs["data-fui-rpc-method"] = "POST"
		carrierAttrs["data-fui-rpc-trigger"] = "input"
		carrierAttrs["data-fui-rpc-debounce-ms"] = strconv.Itoa(debounce)
		carrierAttrs["data-fui-rpc-signal"] = p.Island.Signal
		carrierAttrs["data-hui-combobox-loading"] = w.ComboboxLoading
		Mark(carrierAttrs, "data-hui-combobox-loader")
	}

	listboxAttrs := Attrs(map[string]string{
		"id":         listboxID,
		"role":       "listbox",
		"aria-label": scrubControlBytes(p.Label) + " " + w.ComboboxResultsLabel,
	})
	Mark(listboxAttrs, "data-hui-combobox-listbox")
	// The module announces the result count into the status region
	// through this sentence — on the island listbox after each swap,
	// on the static listbox after each filter.
	listboxAttrs["data-hui-combobox-count"] = w.ComboboxResultCount
	if p.Island != nil && !hasStatic {
		listboxAttrs["data-fui-signal"] = p.Island.Signal
		listboxAttrs["data-fui-signal-mode"] = "html"
		Mark(listboxAttrs, "hidden")
	}
	var rows []render.HTML
	if hasStatic {
		Mark(listboxAttrs, "data-hui-combobox-static")
		Mark(listboxAttrs, "hidden")
		for _, o := range p.Options {
			rows = append(rows, comboboxOptionEl(b, o))
		}
	}

	statusAttrs := Attrs(map[string]string{"role": "status"})
	Mark(statusAttrs, "data-hui-combobox-status")
	statusAttrs["data-hui-combobox-no-results"] = w.ComboboxNoResults
	status := b.El("span", PartComboboxStatus, statusAttrs, render.HTML(""))

	body := []render.HTML{
		b.El("label", PartLabel, Attrs(map[string]string{"for": p.ID}),
			render.Text(scrubControlBytes(p.Label))),
		b.El("div", PartComboboxForm, carrierAttrs,
			b.El("input", PartComboboxInput, inputAttrs)),
		b.El("ul", PartComboboxListbox, listboxAttrs, rows...),
		status,
	}

	rootAttrs := Merge(Safe(p.ExtraAttrs, "role"), nil)
	Mark(rootAttrs, "data-hui-combobox")
	inner := b.El("div", PartRoot, rootAttrs, body...)
	if p.NoScriptAction != "" {
		// The no-script GET: the same query, submitted as a form to the
		// same-origin destination. With script, the submit never fires
		// (Enter picks the active option instead).
		form := b.El("form", PartComboboxForm, Attrs(map[string]string{
			"action": p.NoScriptAction,
			"method": "GET",
			"role":   "none",
		}), inner)
		return form
	}
	return inner
}

// comboboxOptionEl renders one static option row.
func comboboxOptionEl(b Box, o ComboboxOption) render.HTML {
	attrs := Attrs(map[string]string{
		"id": o.ID,
	})
	attrs["role"] = "option"
	if o.Value == "" {
		o.Value = o.Label
	}
	attrs["data-value"] = scrubControlBytes(o.Value)
	if o.Disabled {
		attrs["aria-disabled"] = "true"
	}
	if href := urlsafe.Clean(o.Href, urlsafe.Anchor); href != "" {
		// The module hands the pick to the SPA navigator; an unsafe
		// href drops the navigation affordance entirely.
		attrs["data-fui-push-state"] = href
	}
	kids := []render.HTML{b.El("span", PartText, nil, render.Text(scrubControlBytes(o.Label)))}
	if o.Meta != "" {
		kids = append(kids, b.El("span", PartComboboxOption, nil, render.Text(scrubControlBytes(o.Meta))))
	}
	return b.El("li", PartComboboxOption, attrs, kids...)
}

func init() {
	Register(Spec{
		Name: "Combobox",
		Anatomy: []Part{PartRoot, PartLabel, PartComboboxForm, PartComboboxInput,
			PartComboboxListbox, PartComboboxOption, PartComboboxStatus, PartText},
		Hooks: []string{"data-hui-combobox", "data-hui-combobox-input",
			"data-hui-combobox-listbox", "data-hui-combobox-static",
			"data-hui-combobox-loader", "data-hui-combobox-status",
			"data-hui-combobox-count", "data-hui-combobox-no-results",
			"data-hui-combobox-loading"},
		WithParts: func(s Classes, parts Parts) render.HTML {
			return Combobox(ComboboxProps{ID: "q", Name: "q", Label: "Search",
				Options: []ComboboxOption{{Label: "Docs"}, {Label: "Examples"}}, Parts: parts}, s)
		},
		Cases: func(k Kit) []Case {
			s := k.Classes
			return []Case{{
				Name: "a static combobox",
				Why:  "the options are real content the server rendered, the input owns the keyboard, and no script is needed to read the list",
				HTML: Combobox(ComboboxProps{ID: "q", Name: "q", Label: "Search",
					Options: []ComboboxOption{
						{Label: "Docs", Meta: "/docs"},
						{Label: "Examples", Href: "/examples"},
					}}, s),
			}, {
				Name: "an island combobox",
				Why:  "the results region is bound to the island's signal, and the no-script path is the same-origin GET form wrapping the whole combobox",
				HTML: Combobox(ComboboxProps{ID: "site-q", Name: "q", Label: "Search",
					Island:         &Island{Endpoint: "/island/search", Signal: "search"},
					NoScriptAction: "/search"}, s),
			}}
		},
	})
}
